Switch to Upstream Grafana & Prometheus Containers and Update PostgreSQL Operator Metrics Installer

andrewlecuyer opened the pull request:
With this commit, the PostgreSQL Operator Metrics infrastructure will no longer utilize the `crunchy-prometheus` and `crunchy grafana` containers, and will instead utilize the official upstream Prometheus and Grafana container images as are available on Docker Hub:

- https://hub.docker.com/r/prom/prometheus
- https://hub.docker.com/r/grafana/grafana

Additionally, if RedHat certified containers are needed for Prometheus and/or Grafana, the `ose-prometheus` and `ose-grafana` containers can also be utilized, as are available in the RedHat certified container catalog (see https://catalog.redhat.com/software/containers/explore).

In support of the transition to the upstream Prometheus and Grafana  containers, the installers for the PostgreSQL Operator Metrics infrastructure have been fully separated from the PostgreSQL Operator installers.  This includes the separation of the installers themselves within the PostgreSQL Operator repository - with PostgreSQL Operator Metrics now having it's own Ansible, kubectl and Helm installers - as well as separation of the installation instructions for the metrics infrastructure across these various installation methods within the PostgreSQL Operator documentation.

To support customization of the Prometheus and Grafana applications comprising the PostgreSQL Operator Metrics infrastructure,  ConfigMaps containing custom Prometheus and/or Grafana configuration files can now be created and specified\ during installation.  The configuration settings for supplying these custom configs are as follows:

- **`prometheus_custom_config`:** provide a custom prometheus.yml config
- **`grafana_dashboards_custom_config:`** provide custom Grafana dashboards
- **`grafana_datasources_custom_config:`** provide custom Grafana datasources

Additionally, any dashboards provided by pgmonitor can be enabled/disabled using the **`grafana_dashboards`** configuration setting (a subset of the pgmonitor dashboards are enabled by default, as defined within the metrics infrastructure installation documentation).

Like the PostgreSQL Operator kubectl and Helm installers, the PostgreSQL Operator Metrics kubectl and Helm installers leverage the `pgo-deployer` container to run the PostgreSQL Operator Metrics Ansible installer in order to deploy the metrics infrastructure via a Kubernetes job.  Therefore, the `pgo-deployer` container has now been updated to include both the Operator and Metrics Ansible installers.  Additionally, the `config_sync.sh` script has been updated to sync the kubectl and Helm installers for the PostgreSQL Operator Metrics infrastructure with a master configuration file contained within the PostgreSQL Operator Metrics Ansible installer.

And finally, with this transition to the upstream Prometheus and Grafana containers, the PostgreSQL Operator Metrics Ansible installer has also been updated for consistency and clarity, while also leveraging native configuration settings and capabilities as provided by the upstream containers, as well as directly leveraging pgmonitor resources, wherever possible.  However, certain documentation and naming conventions (some of which are configurable), as well as implementation/design details, have been maintained to facilitate upgrades for current metrics installations and align with the PostgreSQL Operator installers.

**Checklist:**

 <!--- Make sure your PR is documented and tested before submission. Put an `x` in all the boxes that apply: -->
 - [x] Have you added an explanation of what your changes do and why you'd like them to be included?
 - [x] Have you updated or added documentation for the change, as applicable?
 - [x] Have you tested your changes on all related environments with successful results, as applicable?



**Type of Changes:**

 <!--- What types of changes does your code introduce? Put an `x` in all the boxes that apply: -->
 - [ ] Bug fix (non-breaking change which fixes an issue)
 - [x] New feature (non-breaking change which adds functionality)
 - [ ] Breaking change (fix or feature that would cause existing functionality to change)



**What is the current behavior? (link to any open issues here)**

The PostgreSQL Operator Metrics infrastructure utilizes the `crunchy-promtheus` and `crunchy-grafana` containers.

**What is the new behavior (if this is a feature change)?**

The PostgreSQL Operator Metrics infrastructure utilizes the upstream Prometheus and Grafana containers.

[ch8741]

**Other information**:

N/A

jmckulk commented on installers/metrics/ansible/roles/pgo-metrics/tasks/main.yml:
```diff
@@ -0,0 +1,106 @@
+        url: https://github.com/CrunchyData/pgmonitor/archive/{{ pgmonitor_version }}.tar.gz
+        dest: "{{ metrics_dir }}"
+        mode: "0600"
+      tags:
```
tags are defined for the block so I don't think this will be used

andrewlecuyer replied:
These tags have been removed.

jmckulk commented on installers/metrics/ansible/roles/pgo-metrics/tasks/prometheus.yml:
```diff
@@ -0,0 +1,90 @@
+    - name: Set Prometheus Output Directory Fact
+      set_fact:
+        prom_output_dir: "{{ metrics_dir }}/output/prom"
+      tags: always
```
Since these `tags: always` are in a block with `tags: install, update` will they be run for an `uninstall`? Should this always tag be here?

andrewlecuyer replied:
Removed `tags: always` for the directory tasks in both `prometheus.yml` and `grafana.yml`.

jmckulk commented on installers/metrics/ansible/README.md:
```diff
@@ -0,0 +1,15 @@
+This repository contains Ansible Roles for deploying the metrics stack for the
+Crunchy PostgreSQL Operator.
+
+See the [official Crunchy PostgreSQL Operator documentation](https://crunchydata.github.io/postgres-operator/stable/)
```
Would this need to be the `access.crunchydata.com` link?

andrewlecuyer replied:
Switched to the `access.crunchydata.com` link.

jkatz commented on docs/content/installation/metrics/_index.md:
```diff
@@ -0,0 +1,22 @@
+---
+title: "PostgreSQL Operator Metrics"
```
Let's all this "PostgreSQL Operator Monitoring" or "PostgreSQL Operator - Monitoring with pgMonitor"

andrewlecuyer replied:
Just to clarify, do we want to switch to "PostgreSQL Operator Monitoring" across the board, i.e. anywhere "PostgreSQL Operator Metrics" is referenced.  "PostgreSQL Operator Metrics" is scattered all over throughout the docs, so just wanted to verify before making this change.

andrewlecuyer replied:
I went ahead and made a wholesale change to the docs (and a couple of other files) from "PostgreSQL Operator Metrics" to "PostgreSQL Operator Monitoring" (in addition to a few other general changes from "metrics" to "monitoring").  All renames are included in [this commit](https://github.com/CrunchyData/postgres-operator/pull/1757/commits/e476821b3cc6c8fefcb9d2108bbaa51e2c31e1bd).  Also, please note that various directories, files and certain places in the docs (where applicable) as included in this PR still specify "metrics".

jkatz commented on docs/content/installation/metrics/metrics-configuration.md:
```diff
@@ -0,0 +1,123 @@
+
+# PostgreSQL Operator Metrics Installer Configuration
+
+When installing the PostgreSQL Operator Metrics infrastructure you have various configuration options available, which
```
Similar to earlier comment.I think we should ensure we call it "pgMonitor" as we are using that stack.

Also give intros to pgMonitor as needed, explain how it is a one-stop solution for monitoring all of the critical metrics to ensure you run a healthy PostgreSQL cluster.

andrewlecuyer replied:
Just wanted to mention that there is a separate story to create an architecture section for the metrics infrastructure, which will provide a detailed overview of it's various components, e.g. pgmonitor.

However, I can definitely provide an intro here as well.

andrewlecuyer replied:
I went ahead and added an intro to the "PostgreSQL Operator Monitoring" installation page, which is included in [this commit](https://github.com/CrunchyData/postgres-operator/pull/1757/commits/f178310e7979fffeec4fbc08d0346ee7cfcadcba).

andrewlecuyer commented:
Thanks @jkatz & @jmckulk, really appreciate the reviews.

@jkatz yes, Alert Manager is being tracked in a separate story, and will therefore be submitted in a separate PR.

andrewlecuyer commented on installers/metrics/ansible/roles/pgo-metrics/tasks/cleanup.yml:
```diff
@@ -0,0 +1,64 @@
+    - name: Delete Prometheus ConfigMap
+      command: "{{ kubectl_or_oc }} delete configmap crunchy-prometheus -n {{ metrics_namespace }}"
+      ignore_errors: yes
+  tags:
```
All tags for blocks have been moved to the top.

andrewlecuyer commented on docs/content/installation/metrics/other/ansible/updating-metrics.md:
````diff
@@ -0,0 +1,116 @@
+ansible-playbook -i /path/to/inventory.yaml --tags=update --ask-become-pass main.yml
+```
+
+## Verifying the Update
````
Not too worried about a bit of duplication here (these sections shouldn't be too much of a moving target), so will leave as it for this iteration.

andrewlecuyer commented:
Just a note that in addition to addressing the feedback above, I pushed an [additional commit](https://github.com/CrunchyData/postgres-operator/pull/1757/commits/8aa11f9e62efc19f8f96d5450b6241e110236ead) that sets the `runAsUser` to `2` in the `securityContext` for the Grafana and Prometheus pods for non-OpenShift users (i.e. when `disable_fsgroups` is `false`).  This is to ensure the upstream Grafana and Prometheus containers can access data stored from previous versions of the metrics infrastructue, which utilized the `crunchy-grafana` and `crunchy-prometheus` containers that ran as `UID 2`.  This is not an issue in OpenShift, since OpenShift consistently adds the same arbitrary UID regardless of the specific container image being utilized.

cbandy commented on installers/metrics/ansible/roles/pgo-metrics-preflight/tasks/preflight-prometheus.yml:
```diff
@@ -0,0 +1,17 @@
+  tags: always
+  when: lookup('vars', item, default='') == ''
+  loop:
+    - db_port
+    - pgbadgerport
```
I see why these are needed at this time. We should probably (soon) move toward scraping known ports rather than ignoring known ports (that is, `keep` rather than `drop` in prometheus config).

andrewlecuyer replied:
Agreed.  This will take coordination with the pgmonitor team since they now own/control the [default scrape configs](https://github.com/CrunchyData/pgmonitor/blob/9697c0d41e2a353f2dda4c353b272e8be570a935/prometheus/crunchy-prometheus.yml.containers).

cbandy commented on installers/metrics/ansible/roles/pgo-metrics/tasks/cleanup.yml:
```diff
@@ -0,0 +1,64 @@
+    - uninstall-metrics
+  block:
+    - name: Delete Prometheus Deployment
+      command: "{{ kubectl_or_oc }} delete deployment crunchy-prometheus -n {{ metrics_namespace }}"
+      ignore_errors: yes
```
For these `delete` calls, we could use `--ignore-not-found=true` and not ignore errors.

andrewlecuyer replied:
The thought here is to continue on any error, not only "not found" errors.

andrewlecuyer replied:
Adding `--ignore-not-found=true` could def avoid polluting our Ansible logs though ... def something to consider for the future (for metrics and operator installers).

jkatz commented on installers/metrics/ansible/roles/pgo-metrics/templates/prometheus-deployment.json.j2:
```diff
@@ -26,14 +29,15 @@
 {% if not (disable_fsgroup | default(false) | bool) %}
                     {% if (prometheus_supplemental_groups | default('')) != '' %},{% endif -%}
-                    "fsGroup": 26
+                    "fsGroup": 26,
```
Do we need `fsGroup` here? AFAIK this is not being run/owned by the `postgres` user.

andrewlecuyer replied:
We still need `fsGroup` when using cloud storage, e.g. when running in GKE.  I stuck with `26` since this is what we were using with our old `crunchy` metrics containers and installers and wanted to ensure consistency:

https://github.com/CrunchyData/postgres-operator/blob/6fbd804b57ab68cd4286473ce22b2b1ccf58c25e/installers/ansible/roles/pgo-metrics/templates/grafana-deployment.json.j2#L29

Since Kubernetes adds the user running the container to the `fsGroup` specified, and then modifies the group ownership of the files with that same group, this allows proper access to the volumes, even though there is a different user running the container (so with this PR `fsGroups` and users essentially remain consistent with how they previously were with the old metrics containers & installers).

jkatz replied:
OK.

cbandy commented on installers/metrics/ansible/roles/pgo-metrics/files/prometheus.yml:
```diff
@@ -0,0 +1,63 @@
+    target_label: kubernetes_namespace
+  - source_labels: [__meta_kubernetes_pod_name]
+    target_label: pod
+  - source_labels: [__meta_kubernetes_pod_name]
+    regex: (^[^-]*).*
+    target_label: instance
+    replacement: '$1'
```
Is this capturing everything leading up to (and not including) the first dash as `instance`?

andrewlecuyer replied:
This is simply a placeholder copy of the `prometheus.yml` configuration file.  Now that it's been included in pgmonitor itself, we will be grabbing it directly from the pgmonitor project (this change will ocurr with the implementation of CH 8765):

https://github.com/CrunchyData/pgmonitor/blob/9697c0d41e2a353f2dda4c353b272e8be570a935/prometheus/crunchy-prometheus.yml.containers

Therefore, we should address any questions/concerns regarding the default Prometheus config with the pgmonitor team.

cbandy commented on installers/metrics/ansible/roles/pgo-metrics/files/prometheus.yml:
```diff
@@ -0,0 +1,63 @@
+  - source_labels: [__meta_kubernetes_pod_label_role]
+    target_label: role
+    replacement: '$1'
+  - source_labels: [dbname]
+    target_label: dbname
+    replacement: '$1'
+  - source_labels: [relname]
+    target_label: relname
+    replacement: '$1'
+  - source_labels: [schemaname]
+    target_label: schemaname
+    replacement: '$1'
```
These three look like no-ops. Perhaps I'm misunderstanding?

andrewlecuyer replied:
See comment above regarding the containers `prometheus.yaml` config that is now part of pgmonitor (where we will be grabbing it from moving forward).

andrewlecuyer commented on docs/content/installation/metrics/_index.md:
```diff
@@ -0,0 +1,48 @@
+
+The PostgreSQL Operator Monitoring infrastructure is a fully integrated solution for monitoring
+and visualizing metrics captured from PostgreSQL clusters created using the PostgreSQL Operator.
+By leveraging [pgmonitor](https://github.com/CrunchyData/pgmonitor) to configure and integrate
```
Updated per recommendation.
